### DISCUSSION TITLE
An easy way to proxy analytics script on LiteSpeed webservers.

### DISCUSSION THREAD

**Discussion by oldrup** [plausible/analytics#1974] [Anything Else]
Hey

Just wanted to share what I believe is the easiest way to "proxy the script for analytics" when using **LiteSpeed** webservers, which are super common on the WordPress hosts I'm using..

In the Litespeed > Page Optimization menu, there is a rather new tab called "Localize Resources".

Just enter the url for the tracking script, flush the cache, and off you go..

![image](https://user-images.githubusercontent.com/17141205/175821516-c7f2ee86-ee00-4443-9a98-8c9fe41d5b9c.png)

More info on the feature is available here: https://docs.litespeedtech.com/lscache/lscwp/pageopt/#localize-resources

Here's how the script is server from my domain (dev console, network tab)
![image](https://user-images.githubusercontent.com/17141205/175821920-e037a429-0d2b-4c4e-8b5c-49f7c3ae1b3c.png)

My Adblocker (ublock origin) no longer has anything to block.

I tested that visits are actually being tracked, in Firefox and Edge.

Method currently being tested on https://oldrup.net/ - but I'll implement it on my other sites anytime.

Of course inspired by the article at https://plausible.io/docs/proxy/introduction

Will happily provide prettier screenshots for documentation, if someone verifies this approach as working great.

Bjarne

**Comment by metmarkosaric** [comment]
thanks for the feedback @oldrup! i just checked and seems to be working all fine. my ublock origin doesn't catch it. i can still see a call to data-api='https://plausible.io/api/event and our instructions proxy that call too but adblockers don't seem to be touching it in this case.

**Comment by lightson** [comment]
I had a little trouble with this, in that tracking appeared to stop immediately after activiating LS Cache localisation.

I noticed I couldn't see the call to data-api="https://plausible.io/api/event", but then it wasn't in my script src:

`<script defer data-domain="mydomain.com" src="https://plausible.io/js/plausible.js"></script>`

I manually added added data-api, and all is well and recording properly

`<script defer data-api="https://plausible.io/api/event" data-domain="mydomain.com" src="https://plausible.io/js/plausible.js"></script>
`

**Reply by oldrup** [reply]
> I manually added added data-api, and all is well and recording properly

Hey lightson, thank you for your feedback. Are you adding the script to your installation manually?

On my WordPress sites I use the [Plausible plugin on WordPress.org](https://wordpress.org/plugins/plausible-analytics/) as it generates the script just fine on my sites.

The resulting code, with LS Casche Localization being:
`<link rel='dns-prefetch' href='//plausible.io' />
<script async defer data-domain='example.net' data-api='https://plausible.io/api/event' src='https://example.net/wp-content/litespeed/localres/1dGJvdW5kLWxpbmtzLmpz?ver=1.2.4' id='plausible-analytics-js'></script>
<script   id='plausible-analytics-js-after'>
window.plausible = window.plausible || function() { (window.plausible.q = window.plausible.q || []).push(arguments) }
</script>
`
Console reports this:
![image](https://user-images.githubusercontent.com/17141205/179707741-8d131e0a-902c-42e4-aeb3-5436633ff526.png)

And realtime view in Plausible registers my activity:

![image](https://user-images.githubusercontent.com/17141205/179708403-362c2bd8-0405-499d-8b22-6509d04a89c6.png)

Again, with adblocker enabled, which typically would block the analytics-scripts.

![image](https://user-images.githubusercontent.com/17141205/179708935-219f0fab-ec4b-448b-becc-35824a27d254.png)

**Comment by lightson** [comment]
That's correct, I try to keep my use of plugins to a minimum so used the script info provided within plausible settings:

<img width="837" alt="image" src="https://user-images.githubusercontent.com/951483/179712293-9a166e13-e35e-488a-8383-19eca25485a8.png">

LS Cache picks this up fine, but as I said I needed to take the additional step of adding `data-api="https://plausible.io/api/event"` to the code. Not a huge problem, just thought it'd be helpful info to tag on here as it had me confused for a short while. Resulting script as below:

<img width="1068" alt="image" src="https://user-images.githubusercontent.com/951483/179713084-35ea04d0-ed8e-4861-a2c3-1201751757ca.png">

**Reply by oldrup** [reply]
@lightson that valuable input. Thank you so much for sharing. It is important to me that the approach I'm suggesting, using LS Cache to proxy the plausible script, works whether you use the Plausible plugin or add the code manually. 

Reducing the amount of plugins in WordPress is good practice. On sites with ongoing development and changes, I find the plugin handy, though, as it allows me easily to disable tracking while testing, to avoid polluting my data with my own visits.

**Comment by luigirockz** [comment]
This looks like to not be working anymore which is unfortunate, not exactly sure when it did stop.

**Reply by oldrup** [reply]
@luigirockz Yeah, to be honest, I can't say when or why this approach failed. I found that the approach described on https://github.com/Neoflow/ReverseProxy-PlausibleAnalytics was quick and easy to implement, and didn't make me rely on the LiteSpeed cache plugin. I have sites hosted on other types of servers, so that method is just more versatile.

**Reply by lightson** [reply]
It's going to be the data-domain I reckon. I stuck with the non-plugin approach I described above and still working fine with ls cache localising it.

**Reply by guillemap** [reply]
Any update on that? It seems @lightson solution works but not completely:
- The good: the script is properly loaded locally, so no external resources are involved in loading the page.
- The bad: when it's executed it sends a POST request to https://plausible.io/api/event. This request gets blocked by ad blockers, so the main purpose of proxying the script (avoid ad blockers) is not accomplished.
- ...The ugly? guess the only way to make it work is by proxying through something similar to https://github.com/Neoflow/ReverseProxy-PlausibleAnalytics, but it didn't work for me on a litespeed server.

**Comment by zenminimalist** [comment]
@oldrup How did you manage to setup a reverse proxy in Litespeed? I can't seem to get it to work.

**Reply by oldrup** [reply]
@zenminimalist - while the litespeed localization was an easy hack, It did have its shortcomings, and I'd also prefer my tracking isn't depending on my caching configuration.

So I ended up, using either the Plausible WordPress plugin, where reverse proxy is a simple checkbox - or - the manual execution of the script as described on https://github.com/jnessier/ReverseProxy-PlausibleAnalytics .. While it does look as a bit of work, it's really not much. Edit a domain name string in a file, upload three files to the server, execute a line of js, again my WP theme or plugin makes that a breeze.

When testing - be sure to _not_ to opt out of tracking in your browser. That usually is the pit I fall into 😂

**Comment by zenminimalist** [comment]
Thanks for replying @oldrup, even though this is an old thread. ☺️

Actually, I'm running a VPS (not shared hosting, as mentioned [here](https://github.com/jnessier/ReverseProxy-PlausibleAnalytics)) and had trouble setting up a reverse proxy in OLS with my PA docker installation. My configuration kept giving errors. But I think I finally got it work. Hence my asking.

Running PA locally is also something I want to do. Although I'm not quite sure if I understand the whole thing yet.  

**My setup** 
Intends to use `pa.domain0.com` as my analytics hub (subdomain) from which I want to track i.e. domain0.com, domain1.com, domain2.com etc. (all running from the same VPS). I've read that one should ideally use a different domain so the ad blockers don't block it.

Or does localization mean to take the script (`https://pa.domain0.com/js/script.js`) from my analytics hub `pa.domain0.com` and put in the folders of each domain0.com, domain1.com, domain2.com?
